gui/src/steamshortcutparser.cpp

```diff
@@ -0,0 +1,660 @@
+#include <fstream>
+#include <string>
+#include <crc.h>
+#include <vdfstatemachine.h>
+
+#include <iostream>
+#include <curl/curl.h>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+
+#include "vdf_parser.hpp"
+#if defined(_WIN32)
+#include <direct.h>
+#define mkdir(dir, mode) _mkdir(dir)
+#else
+#include <sys/stat.h>
+#endif
+#include <QEventLoop>
+#include <QFile>
+#include <QNetworkReply>
+#include <QUrl>
+#include <sstream>
+#include <steamshortcutparser.h>
+
+namespace SteamShortcutParser {
+
+    std::vector<char> fileHeader = { 0x00, 0x73, 0x68, 0x6F, 0x72, 0x74, 0x63, 0x75, 0x74, 0x73, 0x00 };
+    std::string controller_layout_workshop_id = "3049833406";
+
+    // Callback function to write data to a file
+    size_t WriteCallback(void* contents, size_t size, size_t nmemb, FILE* stream) {
+        return fwrite(contents, size, nmemb, stream);
+    }
+
+    void createDirectories(ChiakiLog* log, const std::string& path) {
+        size_t pos = 0;
+        while ((pos = path.find_first_of("/\\", pos + 1)) != std::string::npos) {
+            std::string subPath = path.substr(0, pos);
+
+            // Check if the directory already exists
+            struct stat info;
+            if (stat(subPath.c_str(), &info) != 0 || !(info.st_mode & S_IFDIR)) {
+                // If not, create the directory
+                if (mkdir(subPath.c_str(), 0777) != 0) {
+                    CHIAKI_LOGE(log, "Error creating directory: %s", subPath.c_str());
+                    return;
+                }
+                CHIAKI_LOGI(log, "Created directory: %s", subPath.c_str());
+            }
+        }
+    }
+
+    void copyFile(ChiakiLog* log, const std::string& sourcePath, const std::string& destinationPath) {
+        // Open the source file for binary input
+        std::ifstream sourceFile(sourcePath, std::ios::binary);
+
+        if (!sourceFile.is_open()) {
+            CHIAKI_LOGE(log, "Error opening the source file: %s", sourcePath.c_str());
+            return;
+        }
+
+        // Open the destination file for binary output
+        std::ofstream destinationFile(destinationPath, std::ios::binary);
+
+        if (!destinationFile.is_open()) {
+            CHIAKI_LOGE(log, "Error opening the destination file: %s", destinationPath.c_str());
+            sourceFile.close(); // Close the source file
+            return;
+        }
+
+        // Copy the contents of the source file to the destination file
+        destinationFile << sourceFile.rdbuf();
+
+        // Close both files
+        sourceFile.close();
+        destinationFile.close();
+
+        CHIAKI_LOGI(log, "File copied successfully from %s to %s", sourcePath.c_str(), destinationPath.c_str());
+    }
+
+    // Function to download a file from a URL to a destination directory
+    bool downloadFile(ChiakiLog* log, const char* url, const char* destPath) {
```

Comment:
The url should be `QUrl` and then you have `QUrl::isLocalFile`. But if you make real local urls `QUrl::fromLocalFile` then QNetworkAccessManager will handle them just fine same as http urls so you don't need to special-case it.